internal/executor/agent/agent.go

```diff
@@ -0,0 +1,201 @@
+package agent
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/api/types/container"
+	"github.com/docker/docker/api/types/mount"
+	"github.com/docker/docker/api/types/volume"
+	"github.com/docker/docker/client"
+	"io"
+	"io/ioutil"
+	"time"
+)
+
+const (
+	// defaultAgentImage is the image we'll extract the agent from.
+	defaultAgentImage = "gcr.io/cirrus-ci-community/cirrus-ci-agent:v1.1"
+
+	// defaultAgentImagePath is a path to an agent's binary in the defaultAgentImage.
+	defaultAgentImagePath = "/bin/cirrus-ci-agent"
+
+	// DefaultAgentVolumePath is a path to an agent's binary in the volume that GetAgentVolume creates.
+	DefaultAgentVolumePath = "/cirrus-ci-agent"
+)
+
+func cleanupContainer(ctx context.Context, cli *client.Client, containerName string) error {
+	timeout := time.Second
+	err := cli.ContainerStop(ctx, containerName, &timeout)
+	if err != nil {
+		return err
+	}
+
+	err = cli.ContainerRemove(ctx, containerName, types.ContainerRemoveOptions{})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func createAgentVolume(
```

Comment:
Do you have any convention or a coding style document in mind that specifically addresses this? 🤔

It seems that by changing this we're essentially introducing our own coding style which will complicate drive-by contributions (simply because it needs to be communicated via a written document, lint rule or a PR comment).

On top of that, this adds unnecessary cognitive load to the developer when making a change (due to ambiguities in determining which function should follow which), yet adds a little benefit with the advent of modern IDE's that make sure there's no practical difference between any of the approaches.

Reply:
That's just how I always wrote code. Don't remember who taught me that but just check how `GetAgentVolume` ends:

```go
return createAgentVolume(ctx, cli, containerNameTemporary, containerNameFinal, volumeName)
```

It will be very nice for human readability to see `createAgentVolume` right after it. We can investigate if golangci-lint has rule for it and if not we can try to implement it.

Reply:
Another point here is to declare public functions first in the file.

Reply:
See 8b1d285e87f0ba7266803329bac08ace47ce8b49.